include/nbl/builtin/hlsl/member_test_macros.hlsl

```diff
@@ -67,5 +73,94 @@ NBL_GENERATE_MEMBER_TESTER(y)
+#define TYPE_FWD(n) Arg##n
+#define DECLVAL_DECLARE(n) impl::declval<Arg##n>()
+
+#define FOR_EACH0(fn)  
+#define FOR_EACH1(fn) fn(1)
+#define FOR_EACH2(fn) fn(2), FOR_EACH1(fn)
+#define FOR_EACH3(fn) fn(3), FOR_EACH2(fn)
+#define FOR_EACH4(fn) fn(4), FOR_EACH3(fn)
+#define FOR_EACH(fn, n) CAT(FOR_EACH, n)(fn)
+
+#define GENERATE_STATIC_METHOD_TESTER_SPEC0(x) \
+template<class T> \
+struct has_static_method_##x<T, typename make_void<decltype(T::x())>::type> : true_type \
+{ \
+    using return_type = decltype(T::x()); \
+    NBL_CONSTEXPR_STATIC_INLINE uint arg_count = 0; \
+}; 
+
+#define GENERATE_STATIC_METHOD_TESTER_SPEC(x, n) \
+template<class T, FOR_EACH(TYPE_DECLARE, n)> \
+struct has_static_method_##x<T, FOR_EACH(TYPE_FWD, n), typename make_void<decltype(T::x(FOR_EACH(DECLVAL_DECLARE, n)))>::type> : true_type \
+{ \
+    using return_type = decltype(T::x(FOR_EACH(DECLVAL_DECLARE, n))); \
+    NBL_CONSTEXPR_STATIC_INLINE uint arg_count = n; \
+}; 
+
+#define GENERATE_STATIC_METHOD_TESTER(x) \
+template<typename T, FOR_EACH(TYPE_DECLARE_DEFAULT, 4), class=void> \
+struct has_static_method_##x : false_type {}; \
+GENERATE_STATIC_METHOD_TESTER_SPEC0(x)  \
+GENERATE_STATIC_METHOD_TESTER_SPEC(x, 1)  \
+GENERATE_STATIC_METHOD_TESTER_SPEC(x, 2)  \
+GENERATE_STATIC_METHOD_TESTER_SPEC(x, 3)  \
+GENERATE_STATIC_METHOD_TESTER_SPEC(x, 4) 
+
+#define GENERATE_METHOD_TESTER_SPEC0(x) \
+template<class T> \
+struct has_method_##x<T, typename make_void<decltype(impl::declval<T>().x())>::type> : impl::if_2_else_1<impl::has_static_method_##x<T>::value> \
+{ \
+    using return_type = decltype(impl::declval<T>().x()); \
+    NBL_CONSTEXPR_STATIC_INLINE uint arg_count = 0; \
+}; 
+
+#define GENERATE_METHOD_TESTER_SPEC(x, n) \
+template<class T, FOR_EACH(TYPE_DECLARE, n)> \
+struct has_method_##x<T, FOR_EACH(TYPE_FWD, n), typename make_void<decltype(impl::declval<T>().x(FOR_EACH(DECLVAL_DECLARE, n)))>::type> : impl::if_2_else_1<impl::has_static_method_##x<T,FOR_EACH(TYPE_FWD, n)>::value> \
+{ \
+    using return_type = decltype(impl::declval<T>().x(FOR_EACH(DECLVAL_DECLARE, n))); \
+    NBL_CONSTEXPR_STATIC_INLINE uint arg_count = n; \
+}; 
+
+/* 
+    types that are impilicitly convertible to each other mess this check up 
+
+    struct S
+    {
+        void a(int) { return 0;}
+    };
+
+    has_method_a<S, float>::value will be true 
+    since float is implicitly convertible to int and
+    due to how we check function signatures at the moment
+*/
+
+#define GENERATE_METHOD_TESTER(x) \
+namespace nbl { \
+namespace hlsl { \
+namespace impl { GENERATE_STATIC_METHOD_TESTER(x) } \
+template<typename T, FOR_EACH(TYPE_DECLARE_DEFAULT, 4), class=void> \
+struct has_method_##x : false_type {}; \
+GENERATE_METHOD_TESTER_SPEC0(x) \
+GENERATE_METHOD_TESTER_SPEC(x, 1) \
+GENERATE_METHOD_TESTER_SPEC(x, 2) \
+GENERATE_METHOD_TESTER_SPEC(x, 3) \
+GENERATE_METHOD_TESTER_SPEC(x, 4) \
+}}
+
+
+GENERATE_METHOD_TESTER(a)
+GENERATE_METHOD_TESTER(b)
```

Comment:
btw I'd also appreviate a variant of `has_method_` that allows to check for an explicitly known return type